File: app/MMarket/c/src/pgs/operationsPg.c
// Copyright 21-Jan-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "pgs/operationsPg.h"
#include "dmc/cgi.h"
#include "dmc/str.h"
#include "dmc/err.h"
#include "dmc/js.h"
#include "data/models.h"
#include "db/evalsDb.h"
#include "db/quotesTb.h"

char *operationsPg_process (Mchar *mrq) {
  char *rq = cgi_rq_string(mrq, "rq");
  if (str_eq(rq, "idata")) {
    char *model_id = cgi_rq_string(mrq, "modelId");
    char *params_js = ochar_some(mchar_get(mrq, "params"));

    Model *md = models_get(model_id);
    AModelEval *evs = evalsDb_read(model_id)->evals;

    ADouble *params = NULL;
    ModelEval *eval = NULL;
    if (!js_is_null(params_js)) {
      params = aDouble_from_js(params_js);
      /**/int ffind (ModelEval *e) {
      /**/  return aDouble_eq(params, e->params, 0.000001);
      /**/}
      eval = oModelEval_nsome(aModelEval_find(evs, ffind));
      if (!eval) params_js = js_wn(); // Enter in next 'if'
    }
    if (js_is_null(params_js)) { // Warning: Not use 'else'
      /**/int fgreater (ModelEval *e1, ModelEval *e2) {
      /**/  return e1->value < e2->value;
      /**/}
      aModelEval_sort(evs, fgreater);
      eval = *(evs->es);
      params = eval->params;
    }

    Quotes *qs = quotesTb_read();
    Result *result = model_result(md, qs, params);
    Achar *nicks = qs->cos;
    ADouble *last_closes = aADouble_get(
      qs->closes, aADouble_size(qs->closes) - 1
    );


    Mchar *rp = mchar_new();
    mchar_put(rp, "model", model_to_js(md));
    mchar_put(rp, "result", result_to_js(result));
    mchar_put(rp, "eval", modelEval_to_js(eval));
    mchar_put(rp, "nicks", achar_to_js(nicks));
    mchar_put(rp, "lastCloses", aDouble_to_js(last_closes));
    mchar_put(rp, "orders", aOrder_to_js(model_orders(md, qs, params)));
    return cgi_rp(rp);
  } else return FAIL(str_f("Unexpected value for 'rq': %s", rq));
}

File: app/KtMMarket/go/pgs/operationsPg/operationsPg.go
// Copyright 04-Jul-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Operations page.
package operationsPg

import (
	"github.com/dedeme/KtMMarket/data/model"
	"github.com/dedeme/KtMMarket/data/modelEval"
	"github.com/dedeme/KtMMarket/data/order"
	"github.com/dedeme/KtMMarket/data/result"
	"github.com/dedeme/KtMMarket/db"
	"github.com/dedeme/KtMMarket/fns"
	"github.com/dedeme/ktlib/arr"
	"github.com/dedeme/ktlib/cgi"
	"github.com/dedeme/ktlib/js"
)

func Process(ck string, mrq cgi.T) string {
	rq := js.Rs(mrq["rq"])
	switch rq {
	case "idata":
		modelId := js.Rs(mrq["modelId"])
		paramsJs := mrq["params"]
		md := model.FromId(modelId)
		evs := db.EvalsDb(modelId).Read().Evals

		var params []float64
		var eval *modelEval.T
		if !js.IsNull(paramsJs) {
			params = arr.Map(js.Ra(paramsJs), js.Rd)
			var ok bool
			eval, ok = arr.Find(evs, func(ev *modelEval.T) bool {
				return fns.EqParams(ev.Params, params)
			})
			if !ok {
				paramsJs = js.Wn()
			}
		}
		if js.IsNull(paramsJs) { // Warning: Not use 'else'
			arr.Sort(evs, func(e1, e2 *modelEval.T) bool {
				return e1.Hvalue > e2.Hvalue
			})
			eval = evs[0]
			params = eval.Params
		}

		qs := db.QuotesTb().Read()
		orders, historic, _, _, _, _, _, profits := md.Simulation(qs, params)
		rs := result.New(
			historic[len(historic)-1],
			arr.Reduce(profits, 0, func(r, e float64) float64 {
				return r + e
			})/float64(len(profits)),
			float64(order.Sales(orders)),
		)

		return cgi.Rp(ck, cgi.T{
			"model":      model.ToJs(md),
			"result":     result.ToJs(rs),
			"eval":       modelEval.ToJs(eval),
			"nicks":      js.Wa(arr.Map(qs.Cos, js.Ws)),
			"lastCloses": js.Wa(arr.Map(qs.Closes[len(qs.Closes)-1], js.Wd)),
			"orders":     js.Wa(arr.Map(orders, order.ToJs)),
		})
	default:
		panic("Value of rq (" + rq + ") is not valid")
	}
}
